include/zephyr/sys/time_units.h

```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
  *    round_up  - Return the ceiling of the resulting fraction
  *    round_off - Return the nearest value to the resulting fraction
  *                (pass both round_up/off as false to get "round_down")
+ *
+ * All of this must be implemented as expressions so that, when constant,
+ * the results may be used to initialize global variables.
  */
-static TIME_CONSTEXPR ALWAYS_INLINE uint64_t z_tmcvt(uint64_t t, uint32_t from_hz,
-						  uint32_t to_hz, bool const_hz,
-						  bool result32, bool round_up,
-						  bool round_off)
-{
-	bool mul_ratio = const_hz &&
-		(to_hz > from_hz) && ((to_hz % from_hz) == 0U);
-	bool div_ratio = const_hz &&
-		(from_hz > to_hz) && ((from_hz % to_hz) == 0U);
-
-	if (from_hz == to_hz) {
-		return result32 ? ((uint32_t)t) : t;
-	}
-
-	uint64_t off = 0;
-
-	if (!mul_ratio) {
-		uint32_t rdivisor = div_ratio ? (from_hz / to_hz) : from_hz;
-
-		if (round_up) {
-			off = rdivisor - 1U;
-		}
-		if (round_off) {
-			off = rdivisor / 2U;
-		}
-	}
-
-	/* Select (at build time!) between three different expressions for
-	 * the same mathematical relationship, each expressed with and
-	 * without truncation to 32 bits (I couldn't find a way to make
-	 * the compiler correctly guess at the 32 bit result otherwise).
-	 */
-	if (div_ratio) {
-		t += off;
-		if (result32 && (t < BIT64(32))) {
-			return ((uint32_t)t) / (from_hz / to_hz);
-		} else {
-			return t / ((uint64_t)from_hz / to_hz);
-		}
-	} else if (mul_ratio) {
-		if (result32) {
-			return ((uint32_t)t) * (to_hz / from_hz);
-		} else {
-			return t * ((uint64_t)to_hz / from_hz);
-		}
-	} else {
-		if (result32) {
-			return (uint32_t)((t * to_hz + off) / from_hz);
-		} else if (const_hz && Z_TMCVT_USE_FAST_ALGO(from_hz, to_hz)) {
-			/* Faster algorithm but source is first multiplied by target frequency
-			 * and it can overflow even though final result would not overflow.
-			 * Kconfig option shall prevent use of this algorithm when there is a
-			 * risk of overflow.
-			 */
-			return ((t * to_hz + off) / from_hz);
-		} else {
-			/* Slower algorithm but input is first divided before being multiplied
-			 * which prevents overflow of intermediate value.
-			 */
-			return (t / from_hz) * to_hz + ((t % from_hz) * to_hz + off) / from_hz;
-		}
-	}
-}
+
+/* true if the conversion is the identity */
+#define z_tmcvt_identity(__from_hz, __to_hz) \
```

Comment:
See above for some more verbose doxygen comments.

 As to types, the general approach I took was to always cast the incoming time value to uint64_t as that matches the original inline parameter type and should end up generating equivalent results.

Splitting the macros into 32/64 bit versions and avoiding a ?: with (uint32_t) on one leg and (uint64_t) on the other avoids making the overall result be uint64_t. I do have a replacement for the original `z_tmcvt` inline  which returned uint64_t, but that's only present to make a couple of tests work without changes. Maybe this should be removed though? It could easily be mis-used and cause 64-bit values to propagate beyond the time conversion operation.

As to generalizing these macros for other uses, that's certainly possible. We should find another use before attempting to do that as the results will end up a lot cleaner.